drivers/xen/console/console.c

```diff
@@ -254,4 +239,59 @@ void hv_console_init(void)
 	console_ready = 1; /* enable notification of backend */
 	/* flush queued output */
 	notify_remote_via_evtchn(console_evtchn);
+	return 0;
+}
+
+__ssz console_out(struct uk_console *dev __unused, const char *str, __sz len)
+{
+	UK_ASSERT(len <= __I_MAX);
+
+	if (unlikely(len == 0))
+		len = strnlen(str, len);
+
+	return hv_console_output(str, len);
+}
+
+__ssz console_in(struct uk_console *dev __unused, char *str __maybe_unused,
+		 __sz maxlen __maybe_unused)
+{
+	UK_ASSERT(maxlen <= __I_MAX);
+	return hv_console_input(str, maxlen);
 }
+
+static struct uk_console_ops console_ops = {
+	.out = console_out,
+	.in = console_in
+};
+
+static struct uk_console console_dev;
+
+#ifdef CONFIG_PARAVIRT
+static int hv_console_prepare(struct ukplat_bootinfo *bi __unused)
+{
+	console_ring = mfn_to_virt(HYPERVISOR_start_info->console.domU.mfn);
+	console_evtchn = HYPERVISOR_start_info->console.domU.evtchn;
+	uk_console_init(&console_dev, "Xen", &console_ops,
+			UK_CONSOLE_FLAG_STDOUT | UK_CONSOLE_FLAG_STDIN);
+	uk_console_register(&console_dev);
+	return 0;
+}
+#endif
+
+#if defined(__aarch64__)
+static int hv_console_prepare(struct ukplat_bootinfo *bi __unused)
+{
+	console_ring =
+	    (struct xencons_interface *)HYPERVISOR_start_info->console.domU.mfn;
+	console_evtchn = HYPERVISOR_start_info->console.domU.evtchn;
+	uk_console_init(&console_dev, "Xen", &console_ops,
+			UK_CONSOLE_FLAG_STDOUT | UK_CONSOLE_FLAG_STDIN);
+	uk_console_register(&console_dev);
+	return 0;
+}
+#endif
+
+UK_BOOT_EARLYTAB_ENTRY(hv_console_prepare, UK_PRIO_EARLIEST);
```

Comment:
Just a note for the future, that anything registered to `UK_PRIO_EARLIEST` is not guaranteed to have access to the cmdline as that is also parsed at the same priority level. Also this function depends on `events_init()` which should eventually also move to the earlytab before `hv_console_prepare()`, and I'm not sure whether these also have dependencies to traps or shared_info. So for now it's probably okay as is, but I anticipate that it will be rearranged later as more things move into earlytab.

Reply:
Good point